### ISSUE TITLE
copy module should preserve mode of src file when mode is not explicitly set

### ISSUE THREAD

**Issue by eikef** [ansible/ansible-modules-core#1124]
##### Summary:

When using the copy module to copy multiple files to their destination, it would be really useful if the mode of the source was preserved. This would allow copying several files in a list without explicitly setting modes for each one.

In any case, this is the expected behavior when going by the documentation (which does not specify a default for mode -- leading one to assume the mode would be preserved unless explicitly specified).
##### Steps To Reproduce:

```
- name: copy test
  copy: src={{ item }} dest=/destination owner=root group=root
  with_items:
    - test1.sh
    - test1.conf
```

with test1.sh locally with 0755 permissions and test1.conf locally with 0644 permissions
##### Expected Results:

test1.sh on target with 0755 permissions and test1.conf on target with 0644 permissions
##### Actual Results:

test1.sh and test1.conf on target with 0644 permissions.

If you were to want to maintain the current behavior for backwards compatibility, the default mode should at least be documented (with possibly a recommendation to look at the synchronize module instead). Alternatively this could be done with something like 

```
- name: copy test
  copy: src={{ item }} dest=/destination owner=root group=root mode=preserve
  with_items:
    - test1.sh
    - test1.conf
```

**Comment by wufeiqun** [comment]
I was also meet with this problem.hoping for a update.

**Comment by semifocused** [comment]
+1 regarding expected behavior and utility of a fix.  Just ran into this as well...

**Comment by ramiel** [comment]
I have this problem too.
+1

**Comment by meersau** [comment]
+1

**Comment by heyongli** [comment]
+1

**Comment by junho85** [comment]
+1

**Comment by flaugher** [comment]
+1

**Comment by rpettersen** [comment]
+1

**Comment by bla-k** [comment]
+1 definetly

**Comment by ProgressiveAstronaut** [comment]
+1

**Comment by bonovoxly** [comment]
+1

**Comment by pabelanger** [comment]
This would be nice to have.

**Comment by robert-blankenship** [comment]
Yes please +1. I'm partially motivated by a lack of sufficient functionality in the `synchronize` module (https://github.com/ansible/ansible-modules-core/issues/2156)

**Comment by tsailiming** [comment]
+1. Too many problems in the synchronize module, like the sudo tty issue.

**Comment by cmvelo** [comment]
👍  +1

**Comment by lkeijser** [comment]
+1

**Comment by ghostofthecode** [comment]
+1

**Comment by CaptTofu** [comment]
+1 this bit me in the tail big-time

**Comment by mnarusze** [comment]
+1

**Comment by russel-cefalo** [comment]
+1

**Comment by lkeijser** [comment]
This ticket has been open for more than a year. While it's cool that many people give a +1, is something going to be done with this? :-)

**Comment by emanuelis** [comment]
+1

**Comment by KGoreczny** [comment]
+1

**Comment by mdef** [comment]
+1

**Comment by CaptTofu** [comment]
what is the latest on this?

**Comment by frank-carnovale** [comment]
+1
me too, i.e. really surprised that recursive copy did not preserve modes of deeply nested files.  Agree with original suggestion "mode=preserve" or similar.

**Comment by feanorknd** [comment]
+1

**Comment by jbguerraz** [comment]
+1 :)

**Comment by jrivera15** [comment]
+1

**Comment by adangert** [comment]
+1

**Comment by ConanChou** [comment]
+1

**Comment by kipras** [comment]
+1

**Comment by venkatautomates** [comment]
+1

**Comment by lowks** [comment]
+1

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by fortijames** [comment]
+1

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by StephaneBunel** [comment]
+1

**Comment by robcalcroft** [comment]
+1

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by LuckySB** [comment]
+1

**Comment by jctanner** [comment]
@LuckySB the fix from ansible/ansible#17780 should land in the 2.3 release. If you are testing against devel and still see this issue, please open a new bug. This one is closed and is no longer being monitored.

**Comment by anassahmed** [comment]
From the commit code, I'm not sure that's what's happening .. you should explicitly set mode: 'preserve' or the code won't work at all!!

https://github.com/ansible/ansible/commit/5502da3cf81c4b34e7c5500c47cec6e1bcdfbbcb#diff-2ae7ee3819390895333fe91e1df8e365R164

**Comment by ushuz** [comment]
> From the commit code, I'm not sure that's what's happening .. you should explicitly set mode: 'preserve' or the code won't work at all!!
> ansible/ansible@5502da3#diff-2ae7ee3819390895333fe91e1df8e365R164

@jctanner The documentation doesn't mention anything about preserving permissions. http://docs.ansible.com/ansible/latest/copy_module.html

**Comment by dagwieers** [comment]
@ushuz Correct, this appears not to be documented, and what's more, this doesn't even work consistently for other modules (i.e. template does not support it).

**Comment by baracudaz** [comment]
As already mentioned, the feature works unfortunately it is not documented. Setting parameter `mode=preserve` will do trick.

**Comment by jojo221119** [comment]
How can this be documented?
